### PR TITLE
ci: Delete VMs older than 23 hours

### PR DESCRIPTION
**Pull request by nbyavuz**
There could be some leftover VMs. Clear these VMs because of their costs and to get back the space their disks occupy.

It is set to get VMs older than 23 hours instead of 1 day because cron job runs daily, so there could be race conditions.

### CODE REVIEW

File: gcp_delete_old_images_and_vms.py [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+    if delete_vms:
+        print("VMs to be deleted: ", ', '.join(
+            vm['name']for vm in delete_vms))
+
+    # delete old VMs
+    for vm in delete_vms:
+        print('Deleting VM =', vm['name'])
+        delete_cmd = base_cmd + ['delete', vm['name'],
+                                 '--zone', vm['zone'], '--quiet']
+        subprocess.run(delete_cmd, check=True)
```
Comment (anarazel) [inline gcp_delete_old_images_and_vms.py:89]: I think we also need to delete old disks for those VMs?

File: gcp_delete_old_images_and_vms.py [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+
+def main():
+    if 'GCP_PROJECT' not in os.environ:
+        print("GCP_PROJECT not set", file=sys.stderr)
+        sys.exit(1)
+
+    delete_vms()
+    # delete_families()
```
Comment (anarazel) [inline gcp_delete_old_images_and_vms.py:98]: what's that about?

File: gcp_delete_old_images_and_vms.py (hunk 1 above)
Comment (nbyavuz) [reply gcp_delete_old_images_and_vms.py:89]: This command actually deletes VMs and their disks.

File: gcp_delete_old_images_and_vms.py (hunk 2 above)
Comment (nbyavuz) [reply gcp_delete_old_images_and_vms.py:98]: My bad, sorry.

File: gcp_delete_old_images_and_vms.py (hunk 1 above)
Comment (anarazel) [reply gcp_delete_old_images_and_vms.py:91]: I think that's only true for disks explicitly owned by VMs - but the way packer creates disk they are not set up that way (because packer needs to access it after shutting the VM down).

Comment (nbyavuz) [reply gcp_delete_old_images_and_vms.py]: Added a explicit way to deleting remaining disks older than 23 hours

File: gcp_delete_old_images_and_instances.py [hunk 3]
```diff
@@ -0,0 +1,104 @@
+#!/usr/bin/env python3
```
Comment (anarazel) [inline gcp_delete_old_images_and_instances.py:1]: Could you put the rename in a separate commit? Unfortunately it makes the diff hard to read...

File: gcp_delete_old_images_and_instances.py [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+
+import os
+import subprocess
+import sys
+
+
+# Delete images older than 2 weeks, except if an image is the newest image in
+# an image family
+def delete_families():
```
Comment (anarazel) [inline gcp_delete_old_images_and_instances.py:10]: This does *not* delete families. It deletes *older* images in families, but never the last image in a family.

File: gcp_delete_old_images_and_instances.py [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+
+    print("deleting images: ", ', '.join(delete_images))
+
+    # finally delete old images
+    delete_cmd = base_cmd + ['delete', '--quiet'] + delete_images
+    subprocess.run(delete_cmd, check=True)
+
+
+# Delete objects older than 23 hour
+def delete_objects(type: str):
```
Comment (anarazel) [inline gcp_delete_old_images_and_instances.py:60]: I don't think "objects" is a good name - that'd include images, which are handled separately.  Maybe delete_leftover_running()?

File: gcp_delete_old_images_and_instances.py [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+        subprocess.run(delete_cmd, check=True)
+
+
+def main():
+    if 'GCP_PROJECT' not in os.environ:
+        print("GCP_PROJECT not set", file=sys.stderr)
+        sys.exit(1)
+
+    delete_objects('instances')
+    delete_objects('disks')
```
Comment (anarazel) [inline gcp_delete_old_images_and_instances.py:99]: I think we should delete the instances / disks in a separate script. We only run the cleanup task if everything succeeded, but we are more likely to accumulate leftover things if there are failures. 

Also, if there are leftover objects, we might fail running due to resource constraints. So we should remove leftover things *before* we run packer.

**Comment by nbyavuz** [comment]
I didn't revert changes on 'gcp_delete_old_images.py' file, they are mostly styling changes(using pycodestyle). Should I revert these changes too?

File: .cirrus.yml [hunk 7]
```diff
@@ -151,6 +167,8 @@ task:
         VERSION: 7-2
       matrix:
         - name: openbsd-vanilla
+          depends_on:
+           - cleanup-leftover-running
```
Comment (anarazel) [inline .cirrus.yml:171]: Can't we do this depend outside of the matrix, instead of repeating it for each "arm"?

File: gcp_delete_leftover_running.py [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+                               'zone': linesplit[1]})
+
+    if delete_list:
+        print(f"{type} to be deleted: ", ', '.join(
+            disk['name'] for disk in delete_list))
+    else:
+        print(f"no {type} to delete")
+
+    for elem in delete_list:
+        print(f'Deleting {type} =', elem['name'])
```
Comment (anarazel) [inline gcp_delete_leftover_running.py:36]: If you're using f-strings, why not also use them for `elem['name']`?

**Review by anarazel — COMMENTED** [review]
Could you split the change to delete-old-images into a separate PR or at least separate commit?

File: .cirrus.yml (hunk 7 above)
Comment (nbyavuz) [reply .cirrus.yml]: If it is implemented like you said, 'depends_on at outside of matrix' overwrites open/netbsd-postgres task's depends_on. They starts to depend 'cleanup-leftover-running' instead of their vanilla tasks.

**Review by anarazel — APPROVED** [review]
Looks good to me. Have you been able to test that this actually works as-is?

**Comment by nbyavuz** [comment]
> Looks good to me. Have you been able to test that this actually works as-is?

Yes, I set time to 1 second and run it locally:

```sh
$ python3 gcp_delete_leftover_running.py 
instances to be deleted:  instance-1, instance-2
Deleting instances = $instance-1
Deleted [https://www.googleapis.com/compute/v1/projects/pg-ci-images-dev/zones/us-central1-a/instances/instance-1].
Deleting instances = $instance-2
Deleted [https://www.googleapis.com/compute/v1/projects/pg-ci-images-dev/zones/us-central1-a/instances/instance-2].
disks to be deleted:  disk-1
Deleting disks = $disk-1
Deleted [https://www.googleapis.com/compute/v1/projects/pg-ci-images-dev/zones/us-central1-a/disks/disk-1].
```

Just pushed a small fix for removing '$' sign while printing.

**Comment by anarazel** [comment]
> Yes, I set time to 1 second and run it locally:

Cool, then let's merge this once the tests have passed.
